### PR TITLE
Added example on Fast Fourier Transformation

### PR DESCRIPTION
**Pull request by deepaksuresh1411**
## Proposed Changes

  - Added Fast Fourier Transformation example

### CODE REVIEW

File: examples/fast_fourier_transformation/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Copyright (c) 2022 Deepak Suresh
+
+# SPDX-License-Identifier: BSL-1.0
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+if(NOT HPX_WITH_DISTRIBUTED_RUNTIME)
+  return()
+endif()
```
Comment (hkaiser) [inline examples/fast_fourier_transformation/CMakeLists.txt:9]: The example uses local functionalities only (AFAICS), why do you require the distributed runtime to be available?

File: examples/CMakeLists.txt [hunk 2]
```diff
@@ -27,6 +27,7 @@ set(subdirs
     throttle
     tuplespace
     transpose
+    fast_fourier_transformation
```
Comment (hkaiser) [inline examples/CMakeLists.txt:30]: Can you sort this alphabetically, please?

File: examples/fast_fourier_transformation/fft.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+#include <iostream>
+#include <math.h>
+#include <vector>
+
+// For storing complex values of nth roots of unity
+using cd = std::complex<double>;
+
+// Function for fast fourier transformation
+std::vector<cd> fft(const std::vector<cd>& a)
```
Comment (hkaiser) [inline examples/fast_fourier_transformation/fft.cpp:18]: We prefer using 'east const' over 'west const' notations. Could you adapt your code, please?

File: examples/CMakeLists.txt (hunk 2 above)
Comment (deepaksuresh1411) [reply examples/CMakeLists.txt:30]: Sure I'll do it

File: examples/fast_fourier_transformation/CMakeLists.txt (hunk 1 above)
Comment (deepaksuresh1411) [reply examples/fast_fourier_transformation/CMakeLists.txt:9]: Oh! I think it was by mistake I didn't comment out this one

**Comment by deepaksuresh1411** [comment]
@hkaiser should I implement iterative solution rather than recursive one . Please let me know I'll update with that.

**Comment by gonidelis** [comment]
@deepaksuresh1411 Thanks for the useful example! Maybe it's worth providing both kinds of solutions and compare the performance? 

Also please check your CI results and work towards fixing the corresponding errors.

**Comment by deepaksuresh1411** [comment]
@gonidelis yeah sure I'll do it. Thanks

**Comment by dimitraka** [comment]
@deepaksuresh1411 Would you like to provide also some documentation for your example to have a more concrete result? If so, we can discuss that further!

**Comment by deepaksuresh1411** [comment]
> @deepaksuresh1411 Would you like to provide also some documentation for your example to have a more concrete result? If so, we can discuss that further!

Sure! , but might take some time , because of prior commitments .

**Comment by dimitraka** [comment]
> > @deepaksuresh1411 Would you like to provide also some documentation for your example to have a more concrete result? If so, we can discuss that further!
> 
> Sure! , but might take some time , because of prior commitments .

Great! Feel free to ask whatever you need. You can also check the documentation of our examples https://hpx-docs.stellar-group.org/branches/master/html/examples.html to take an idea of how to write it.

**Comment by deepaksuresh1411** [comment]
> @deepaksuresh1411 We are preparing the new release currently. Feel free to polish your PR if you want it to be included in our new set of examples ;)

I would love to get this PR merged, but right now I'm very busy with my college work. And moreover I'm stuck in the CMakeLists.txt can you please look into the problem, if you have time.

**Comment by gonidelis** [comment]
> > @deepaksuresh1411 We are preparing the new release currently. Feel free to polish your PR if you want it to be included in our new set of examples ;)
> 
> I would love to get this PR merged, but right now I'm very busy with my college work. And moreover I'm stuck in the CMakeLists.txt can you please look into the problem, if you have time.

What problem? Could you provide a link?

**Comment by deepaksuresh1411** [comment]
> > > @deepaksuresh1411 We are preparing the new release currently. Feel free to polish your PR if you want it to be included in our new set of examples ;)
> > 
> > 
> > I would love to get this PR merged, but right now I'm very busy with my college work. And moreover I'm stuck in the CMakeLists.txt can you please look into the problem, if you have time.
> 
> What problem? Could you provide a link?

It's the circle CI and failing tests

**Comment by SAtacker** [comment]
> > > > @deepaksuresh1411 We are preparing the new release currently. Feel free to polish your PR if you want it to be included in our new set of examples ;)
> > > 
> > > 
> > > I would love to get this PR merged, but right now I'm very busy with my college work. And moreover I'm stuck in the CMakeLists.txt can you please look into the problem, if you have time.
> > 
> > 
> > What problem? Could you provide a link?
> 
> It's the circle CI and failing tests

I tried it locally and this should work.
```
diff --git a/examples/fast_fourier_transformation/CMakeLists.txt b/examples/fast_fourier_transformation/CMakeLists.txt
index 436cf16e54..76565aafa6 100644
--- a/examples/fast_fourier_transformation/CMakeLists.txt
+++ b/examples/fast_fourier_transformation/CMakeLists.txt
@@ -20,6 +20,6 @@ foreach(example_program ${example_programs})
     FOLDER "Examples/fast_fourier_transformation"
   )
 
-  add_hpx_example_target_dependencies("fft" ${example_program})
+  add_hpx_example_target_dependencies("fast_fourier_transformation" ${example_program})
 
 endforeach()
```

**Comment by deepaksuresh1411** [comment]
> > > > > @deepaksuresh1411 We are preparing the new release currently. Feel free to polish your PR if you want it to be included in our new set of examples ;)
> > > > 
> > > > 
> > > > I would love to get this PR merged, but right now I'm very busy with my college work. And moreover I'm stuck in the CMakeLists.txt can you please look into the problem, if you have time.
> > > 
> > > 
> > > What problem? Could you provide a link?
> > 
> > 
> > It's the circle CI and failing tests
> 
> I tried it locally and this should work.
> 
> ```
> diff --git a/examples/fast_fourier_transformation/CMakeLists.txt b/examples/fast_fourier_transformation/CMakeLists.txt
> index 436cf16e54..76565aafa6 100644
> --- a/examples/fast_fourier_transformation/CMakeLists.txt
> +++ b/examples/fast_fourier_transformation/CMakeLists.txt
> @@ -20,6 +20,6 @@ foreach(example_program ${example_programs})
>      FOLDER "Examples/fast_fourier_transformation"
>    )
>  
> -  add_hpx_example_target_dependencies("fft" ${example_program})
> +  add_hpx_example_target_dependencies("fast_fourier_transformation" ${example_program})
>  
>  endforeach()
> ```
> 
> Output:
> 
> ```
> satacker@ubuntu:/home/extended/hpx/build$ bin/fft 
> (18.6815,33.7669)
> (10.7477,-21.5506)
> (17.3664,-11.8887)
> (-2.79554,-0.327635)
> ```

The program was already running locally, it was just the circle CI was causing some trouble. Might be I'm missing something in the code. BTW thanks for taking interest in this.

**Comment by deepaksuresh1411** [comment]
@gonidelis   The cmake-format was the problem for some failing tests, now it is sorted and fixed. Can you look at the other failing tests, if you have time.

**Comment by gonidelis** [comment]
> @gonidelis The cmake-format was the problem for some failing tests, now it is sorted and fixed. Can you look at the other failing tests, if you have time.

Nice work! All tests seem to pass. Could you please provide the corresponding documentation according to our examples? https://hpx-docs.stellar-group.org/branches/master/html/examples.html

Here you can fine the source code for the docs: https://github.com/STEllAR-GROUP/hpx/tree/master/docs/sphinx/examples

Make sure that you explain the reasoning behind your example, the usefulness and most important describe the HPX features that use and how could these be usefull for future users. It's an example case after all ;)

**Comment by deepaksuresh1411** [comment]
> > @gonidelis The cmake-format was the problem for some failing tests, now it is sorted and fixed. Can you look at the other failing tests, if you have time.
> 
> Nice work! All tests seem to pass. Could you please provide the corresponding documentation according to our examples? https://hpx-docs.stellar-group.org/branches/master/html/examples.html
> 
> Here you can fine the source code for the docs: https://github.com/STEllAR-GROUP/hpx/tree/master/docs/sphinx/examples
> 
> Make sure that you explain the reasoning behind your example, the usefulness and most important describe the HPX features that use and how could these be usefull for future users. It's an example case after all ;)

Sure! :)

**Comment by dimitraka** [comment]
@deepaksuresh1411 Do you have any updates about the documentation part as well? An example is always more complete when it is combined with the proper documentation to be useful for the users!

**Comment by deepaksuresh1411** [comment]
> @deepaksuresh1411 Do you have any updates about the documentation part as well? An example is always more complete when it is combined with the proper documentation to be useful for the users!

TBH I am trying my best, it's just I am collecting some materials for documentation ,as I have never done documentation before and I have to make some changes in code also.
I will also be making a comparison between HPX and OpenMP.

**Comment by dimitraka** [comment]
> > @deepaksuresh1411 Do you have any updates about the documentation part as well? An example is always more complete when it is combined with the proper documentation to be useful for the users!
> 
> TBH I am trying my best, it's just I am collecting some materials for documentation ,as I have never done documentation before and I have to make some changes in code also. I will also be making a comparison between HPX and OpenMP.

If you want any info about the documentation, feel free to ask! There is no extra material needed apart from checking the docs we already have for other examples. The only thing you have to do is basically explain what your code does and to include a short introduction. Thanks again!

**Comment by deepaksuresh1411** [comment]
> 

Yeah sure, I'll do it ASAP, after some code changes

**Comment by hkaiser** [comment]
@deepaksuresh1411 We have finished the V1.8 release now. Could you please rebase your branch onto master and perhaps add the documentation changes you promised to add?

**Comment by deepaksuresh1411** [comment]
> @deepaksuresh1411 We have finished the V1.8 release now. Could you please rebase your branch onto master and perhaps add the documentation changes you promised to add?

Sorry, for the delay. Is this code looks good to you for main codebase of hpx. If that so, then I might add some documentation for this. coz, even I am confused with the fft example. As it should work for any general cases, but its only working for certain inputs.

**Comment by hkaiser** [comment]
> > @deepaksuresh1411 We have finished the V1.8 release now. Could you please rebase your branch onto master and perhaps add the documentation changes you promised to add?
> 
> Sorry, for the delay. Is this code looks good to you for main codebase of hpx. If that so, then I might add some documentation for this. coz, even I am confused with the fft example. As it should work for any general cases, but its only working for certain inputs.

If the example doesn't properly work for all cases - do you plan to fix it? Otherwise I'd suggest closing this PR without merging.

**Comment by deepaksuresh1411** [comment]
> > > @deepaksuresh1411 We have finished the V1.8 release now. Could you please rebase your branch onto master and perhaps add the documentation changes you promised to add?
> > 
> > 
> > Sorry, for the delay. Is this code looks good to you for main codebase of hpx. If that so, then I might add some documentation for this. coz, even I am confused with the fft example. As it should work for any general cases, but its only working for certain inputs.
> 
> If the example doesn't properly work for all cases - do you plan to fix it? Otherwise I'd suggest closing this PR without merging.

Yeah, I will be adding updated example when I am ready with correct algorithm. Until then I think we should close this PR.

**Pull request closed without merging**